src/limb.rs

```diff
@@ -0,0 +1,220 @@
+use {rand, polyfill, c, core, error};
+
+// XXX: Not correct for x32 ABIs.
+#[cfg(target_pointer_width = "64")] pub type Limb = u64;
+#[cfg(target_pointer_width = "32")] pub type Limb = u32;
+#[cfg(target_pointer_width = "64")] pub const LIMB_BITS: usize = 64;
+#[cfg(target_pointer_width = "32")] pub const LIMB_BITS: usize = 32;
+
+pub const LIMB_BYTES: usize = (LIMB_BITS + 7) / 8;
+
+/// References a positive integer range `[1..max_exclusive)` where
+/// `max_exclusive` is a slice of limbs ordered least-significant-limb to
+/// most-significant-limb (and the limbs use the native endianness).
+pub struct Range<'a> {
+    pub max_exclusive: &'a [Limb],
+}
+
+impl <'a>Range<'a> {
+
+    pub fn from_max_exclusive(max_exclusive: &[Limb]) -> Range {
+
+        debug_assert!(!limbs_are_zero_constant_time(max_exclusive));
+        debug_assert!(max_exclusive[max_exclusive.len() - 1] > 0);
+
+        Range {
+            max_exclusive: max_exclusive
+        }
+    }
+
+    fn leading_zero_bits(&self) -> usize {
+        let most_significant_limb =
+            self.max_exclusive[self.max_exclusive.len() - 1];
+
+        for i in 0..(LIMB_BITS - 1) {
+            let bit = 1 << (LIMB_BITS - 1 - i);
+            if most_significant_limb & bit > 0 {
+                return i;
+            }
+        }
+
+        unreachable!();
+    }
+
+    /// Are these little-endian limbs within the range?
+    ///
+    /// Checks in constant time.
+    pub fn are_limbs_within(&self, limbs: &[Limb]) -> bool {
+        assert_eq!(self.max_exclusive.len(), limbs.len());
+
+        let is_gt_zero = !limbs_are_zero_constant_time(limbs);
+        let is_lt_max = limbs_less_than_limbs_constant_time(limbs, self.max_exclusive);
+
+        is_lt_max && is_gt_zero
+    }
+
+    // Loosely based on NSA Guide Appendix B.2:
+    // "Key Pair Generation by Testing Candidates".
+    //
+    // TODO: DRY-up with `ec::suite_b::private_key::generate_private_key`?
+    //
+    /// Chooses a positive integer within the range and stores it into `dest`
+    /// encoded as limbs ordered least-significant-limb to most-significant-limb.
+    ///
+    /// This function is intended to be suitable for generating private keys.
+    pub fn sample_into_limbs(&self, dest: &mut [Limb], rng: &rand::SecureRandom)
+                             -> Result<(), error::Unspecified> {
+
+        assert_eq!(self.max_exclusive.len(), dest.len());
+
+        let bits_to_mask = self.leading_zero_bits();
+
+        // XXX: The value 100 was chosen to match OpenSSL due to uncertainty of
+        // what specific value would be better, but it seems bad to try 100 times.
+        for _ in 0..100 {
+            {
+                let mut dest_as_bytes = limbs_as_bytes_mut(dest);
+                try!(rng.fill(&mut dest_as_bytes));
+            }
+
+            if bits_to_mask > 0 {
+                let mask: Limb = (1 << (LIMB_BITS - bits_to_mask)) - 1;
```

Comment:
Thanks for factoring out the leading zero bits calculation.

I think we should probably follow this up by factoring out the calculation of mask out of the loop:

```
let mask = if bits_to_mask > 0 { 1 << (LIMB_BITS - bits_to_mask)) - 1 } else { LIMB_MAX };
```

Actually, if we look ahead a little bit to how we'll eventually rewrite `BN_mod_inverse_blinded`, we actually need to generate 2 random values in the same range, so it might make the most sense to store the mask as a field of `Range`. It won't make anything faster but it will be easier to see that the mask is purely function of the modulus, not a function of any of the input.